### PR TITLE
Fix TensorFlow Lite for Microcontrollers CMSIS-NN build and add MaxPool implementation

### PR DESCRIPTION
**Pull request by dansitu**
This PR does two things:

- Fixes building of TensorFlow Lite for Microcontrollers examples with CMSIS-NN and mbed, by removing a reference to an undeclared function and by instructing the build scripts to use a newer version of the CMSIS libraries.
- Adds CMSIS-NN optimized implementation of MaxPool to the CMSIS-NN optimized kernels

### CODE REVIEW

**Comment by gbaned** [comment]
@dansitu Thank you for your contribution. Can you please sign CLA? Thanks!

**Comment by dansitu** [comment]
@googlebot I signed it!

**Comment by dansitu** [comment]
@googlebot I signed it!

File: tensorflow/lite/micro/kernels/cmsis-nn/conv.cc [hunk 1]
```diff
@@ -268,9 +268,6 @@ TfLiteStatus EvalQuantizedPerChannel(
              (output_width % 4 == 0) && batches == 1) {
     const int32_t buf_size = arm_convolve_1_x_n_s8_get_buffer_size(
         input_depth, filter_width, filter_height);
-    if (get_cmsis_scratch_buffer(context, &buf, buf_size) != kTfLiteOk) {
-      return kTfLiteError;
-    }
     if (arm_convolve_1_x_n_s8(
```
Comment (freddan80) [inline tensorflow/lite/micro/kernels/cmsis-nn/conv.cc:271]: This is also fixed (along with some other stuff) in https://github.com/tensorflow/tensorflow/pull/38123

File: tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc [hunk 2]
```diff
@@ -206,6 +206,73 @@ void MaxEvalQuantizedUInt8(TfLiteContext* context, TfLiteNode* node,
                          GetTensorData<uint8_t>(output));
 }
 
+TfLiteStatus MaxEvalInt8(TfLiteContext* context, const TfLiteNode* node,
+                             const TfLitePoolParams* params, const OpData* data,
+                             TfLiteTensor* input, TfLiteTensor* output) {
```
Comment (freddan80) [inline tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:211]: make input const

File: tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc [hunk 3]
```diff
@@ -277,7 +344,8 @@ TfLiteStatus MaxEval(TfLiteContext* context, TfLiteNode* node) {
   auto* params = reinterpret_cast<TfLitePoolParams*>(node->builtin_data);
   OpData data;
 
-  const TfLiteTensor* input = GetInput(context, node, kInputTensor);
+  TfLiteTensor* input = &context->tensors[flatbuffers::EndianScalar(
+      node->inputs->data[kInputTensor])];
```
Comment (freddan80) [inline tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:348]: We should be able to keep this const. The background why I introduced this hack for avg pool is that we use a non-const input there (we'll fix it soon). For maxpool, we use a const input. Sorry for the confusion!

File: tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc [hunk 4]
```diff
@@ -166,7 +166,7 @@ TfLiteStatus AverageEvalInt8(TfLiteContext* context, const TfLiteNode* node,
 
 void MaxEvalFloat(TfLiteContext* context, TfLiteNode* node,
                   TfLitePoolParams* params, OpData* data,
-                  const TfLiteTensor* input, TfLiteTensor* output) {
+                  TfLiteTensor* input, TfLiteTensor* output) {
```
Comment (freddan80) [inline tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:169]: make input const

File: tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc [hunk 5]
```diff
@@ -187,7 +187,7 @@ void MaxEvalFloat(TfLiteContext* context, TfLiteNode* node,
 
 void MaxEvalQuantizedUInt8(TfLiteContext* context, TfLiteNode* node,
                            TfLitePoolParams* params, OpData* data,
-                           const TfLiteTensor* input, TfLiteTensor* output) {
+                           TfLiteTensor* input, TfLiteTensor* output) {
```
Comment (freddan80) [inline tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:190]: make input const

**Review by freddan80 — CHANGES_REQUESTED** [review]

**Comment by freddan80** [comment]
Thanks for this contribution @dansitu ! Much appreciated. It looks good in general, I left a few minor comments.

File: tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc (hunk 3 above)
Comment (freddan80) [reply tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:348]: After a second look. The max pooling input is also non-const. I removed the comments. We'll fix that soon.

**Comment by felix-johnny** [comment]
Looks good!

File: tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc (hunk 3 above)
Comment (dansitu) [reply tensorflow/lite/micro/kernels/cmsis-nn/pooling.cc:348]: Sounds good—I played around a bit to see if I could get it to work const and then noticed you had the same issue with avg.

File: tensorflow/lite/micro/kernels/cmsis-nn/conv.cc (hunk 1 above)
Comment (dansitu) [reply tensorflow/lite/micro/kernels/cmsis-nn/conv.cc:271]: SG, I assume it's fine to leave it in here since it'll just merge out in the end.

**Comment by gbaned** [comment]
@dansitu Can you please resolve conflicts? Thanks!

**Comment by dansitu** [comment]
> @dansitu Can you please resolve conflicts? Thanks!

Done!

**Review by petewarden — DISMISSED** [review]

**Comment by dansitu** [comment]
I've also added a fix to `flatbuffer_conversions.cc` where the `ParseOpData` function sometimes did not have a return value, which caused problems with compilers that don't add an implicit return value.

**Review by petewarden — APPROVED** [review]

**Comment by gbaned** [comment]
@mihaimaruseac Can you please take a look on cla/google test failure? Thanks!

**Pull request merged**